MonoGlobalAssemblyCache: faster resolves + resolve facade assemblies

abock opened the pull request:
### Summary

Overall this PR improves assembly resolution in `csi` when running on Mono by resolving facade assemblies like `netstandard.dll`. It also employs a few tricks to reduce resolution time (~4.5x faster according to `Stopwatch` on my machine across 50 iterations before and after).

This PR also addresses an issue where assembly names were and have always been incorrect when running on Mono (see https://github.com/dotnet/roslyn/pull/39369#issuecomment-545970358 below). This issue is exposed in unit tests, but these tests do not run on CI. Locally and as part of this PR, all `GlobalAssemblyCacheTests` now pass on Mono.

### Details

1. Use `AssemblyName.GetAssemblyName(path)` instead of `new AssemblyName(path)`.

2. Don't special case `mscorlib.dll` - we already will have a reference to it, and the changes in (2) allow it to still resolve. Remove `GetCorlibPaths`.

3. Have `GetCacAssemblyPaths` bail early if it finds an assembly that lives alongside `mscorlib.dll`, or in the `Facades` directory. This These assemblies in Mono are considered "core" and will always symlink directly to the GAC. Resolve facade assemblies as well now. The Mono distribution of `csi` has shipped with a response file that explicitly referenced `Facades/netstandard.dll` and `Facades/System.Runtime.dll` and this change obsoletes that. As a consequence, local builds of `csi.exe` now work.

4. Minor perf work around converting a public key token to a string by deferring that conversion until only necessary, and use `byte.ToString` instead of `StringBuilder.AppendFormat`. Also use a `PooledStringBuilder`.

5. `Tuple` → `ValueTuple`

agocke commented:
cc @tmat

tmat commented on src/Compilers/Shared/GlobalAssemblyCacheHelpers/MonoGlobalAssemblyCache.cs:
```diff
@@ -222,5 +199,21 @@ public override IEnumerable<string> GetAssemblySimpleNames(ImmutableArray<Proces
             return assemblyIdentity;
         }
+
+        static string ToHexString(byte[] bytes)
```
>static [](start = 8, length = 6)

nit: missing `private`

tmat commented on src/Compilers/Shared/GlobalAssemblyCacheHelpers/MonoGlobalAssemblyCache.cs:
```diff
@@ -17,54 +17,48 @@ namespace Microsoft.CodeAnalysis
     internal sealed class MonoGlobalAssemblyCache : GlobalAssemblyCache
     {
-        public static readonly ImmutableArray<string> RootLocations;
+        private static readonly string CorlibDirectory;
```
>CorlibDirectory [](start = 39, length = 15)

nit: the convention is to use `s_` prefix

tmat commented:
@abcok Actually, I don't think we have a test that covers resolving the facades from GAC. Would you mind adding on to  https://github.com/dotnet/roslyn/blob/master/src/Scripting/CoreTest.Desktop/GlobalAssemblyCacheTests.cs?

jaredpar commented on src/Compilers/Shared/GlobalAssemblyCacheHelpers/MonoGlobalAssemblyCache.cs:
```diff
@@ -17,54 +17,48 @@ namespace Microsoft.CodeAnalysis
         {
-            RootLocations = ImmutableArray.Create(GetMonoCachePath());
-        }
+            var corlibAssemblyFile = typeof(object).GetTypeInfo().Assembly.Location;
```
Don't believe `GetTypeInfo` is necessary here.

abock replied:
Other uses of `typeof()` to get at the type's assembly are going through `GetTypeInfo`. I also don't believe it should be necessary anymore, but I was trying to stay consistent. No opinion here (the other existing just below this one is `typeof(Uri).GetTypeInfo().Assembly...` to get at `System.dll`. I assume all of these are leftovers from pre-netstandard2.0 days?

abock commented:
@tmat so it appears these tests are not run in CI and the existing test in question that stresses `MonoGlobalAssemblyCache` is already failing locally against the HEAD of master, and it makes sense why:

```csharp
new AssemblyName(path)
```

Here, `path` is the absolute path to the assembly on disk that was found. Mono's `AssemblyName(string)` constructor doesn't validate the argument the same was as core, and ends up such that `new AssemblyName(path).Name == path`.

In any case, I'm not sure how the existing tests ever worked on Mono, because the full path is used as a valid assembly name, yet the tests assert just the filename without extension, e.g. it is expected that the following is true:

```csharp
Assert.Equal("mscorlib", new AssemblyName("/Absolute/Path/To/mscorlib.dll").Name);
```

### .NET Core

```csharp
👉 dotnet exec ./artifacts/bin/csi/Debug/netcoreapp2.1/publish/csi.dll
Microsoft (R) Visual C# Interactive Compiler version 3.4.0-dev ()
Copyright (C) Microsoft Corporation. All rights reserved.

Type "#help" for more information.
> new System.Reflection.AssemblyName("/Full/Path/To/mscorlib.dll").Name
System.IO.FileLoadException: The given assembly name or codebase was invalid. (Exception from HRESULT: 0x80131047)
  + System.Reflection.AssemblyName.nInit(out System.Reflection.RuntimeAssembly, bool)
  + System.Reflection.AssemblyName..ctor(string)
  + <Initialize>.MoveNext()
>
```

### Mono

```csharp
👉 mono ./artifacts/bin/csi/Debug/net472/csi.exe /r:/Library/Frameworks/Mono.framework/Versions/Current/lib/mono/4.5/Facades/netstandard.dll
Microsoft (R) Visual C# Interactive Compiler version 3.4.0-dev ()
Copyright (C) Microsoft Corporation. All rights reserved.

Type "#help" for more information.
> new System.Reflection.AssemblyName("/Full/Path/To/mscorlib.dll").Name
"/Full/Path/To/mscorlib.dll"
>
```

I will fix the existing bad `AssemblyName`s that `MonoGlobalAssemblyCache` is (and has been) yielding and add the test for Facades, but it would be great to actually run these tests in CI.

abock commented:
@tmat tl;dr `MonoGlobalAssemblyCache` has been doing `new AssemblyName(pathToFile)` and should be doing `AssemblyName.GetAssemblyName(pathToFile`).

abock commented:
Also note - at current HEAD of master, there are 9 other failing tests (`Microsoft.CodeAnalysis.Scripting.Hosting.UnitTests.MetadataShadowCopyProviderTests`) that fail on Mono on macOS in the same project (`Scripting/CoreTest.Desktop`).

abock commented:
@tmat fixed the existing tests by fixing assembly name from path issue as detailed above. This is a separate commit from the original intent of this PR.

Added test for resolving netstandard and addressed style nits. @jaredpar removed the two calls to `Type.GetTypeInfo` as well.

abock commented:
Went ahead and fixed `MetadataShadowCopyProviderTests` to run on Mono as well. Only one test I couldn't reason about, so I marked it `[ClrOnlyFact]`. It'd be great to get `Scripting/CoreTest.Desktop` tests running in CI on macOS and Windows at least?

abock commented:
@tmat ping

abock commented:
It seems that Mono may now be using the CoreFX sources for`AssemblyName`, which actually now breaks `csi` in Mono 6.8. This PR should coincidentally address that breakage, so we really should get this merged. 

We now see this exception on Mono 6.8 and this PR should fix that: https://gist.github.com/DavidKarlas/e42af9352afd70552875e0a930e85bd4

sandyarmstrong commented:
@jinujoseph is anything preventing this from merging now? We'd like to get this into Mono 6.8, as our engineers keep running into the `csi` issue mentioned in https://github.com/dotnet/roslyn/pull/39369#issuecomment-550348894 .

tmat commented:
+ @jaredpar Re tests on Mono - all tests are currently [disabled](https://github.com/dotnet/roslyn/blob/master/azure-pipelines.yml#L186) due to https://github.com/mono/mono/issues/16373. The issue is closed though, so perhaps we can update to the latest Mono and enable them now?
